evm/src/cpu/kernel/asm/core/precompiles/expmod.asm

```diff
@@ -1,3 +1,382 @@
+    DUP1
+    %ceil_div_const(16)
+    // stack: num_limbs, num_bytes, len, len, l_M, l_E, l_B, kexit_info
+    SWAP1
+    // stack: num_bytes, num_limbs, len, len, l_M, l_E, l_B, kexit_info
+    DUP7
+    DUP7
+    ADD
+    %add_const(96)
+    // stack: 96 + l_B + l_E, num_bytes, num_limbs, len, len, l_M, l_E, l_B, kexit_info
+    PUSH @SEGMENT_CALLDATA
+    GET_CONTEXT
+    // stack: ctx, @SEGMENT_CALLDATA, 96 + l_B + l_E, num_bytes, num_limbs, len, len, l_M, l_E, l_B, kexit_info
+    %mload_bytes_as_limbs
+    // stack: num_limbs, len, limbs[num_limbs-1], .., limbs[0], len, l_M, l_E, l_B, kexit_info
+    SWAP1
+    %mul_const(2)
+    // stack: m_loc=2*len, num_limbs, limbs[num_limbs-1], .., limbs[0], len, l_M, l_E, l_B, kexit_info
+    %store_limbs
+    // stack: len, l_M, l_E, l_B, kexit_info
+
+    SWAP3
+    %pop3
+    // stack: len, kexit_info
+
+    PUSH expmod_contd
+    // stack: expmod_contd, len, kexit_info
+    DUP2
+    // stack: len, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(11)
+    // stack: s5=11*len, len, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(9)
+    // stack: s4=9*len, len, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, s4, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(7)
+    // stack: s3=7*len, len, s4, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, s3, s4, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(5)
+    // stack: s2=5*len, len, s3, s4, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, s2, s3, s4, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(4)
+    // stack: s1=4*len, len, s2, s3, s4, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(3)
+    // stack: out=3*len, len, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, out, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+
+    DUP1
+    %mul_const(2)
+    // stack: m_loc=2*len, len, out, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+    SWAP1
+    // stack: len, m_loc, out, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+
+    PUSH 0
+    // stack: b_loc=0, e_loc=len, m_loc, out, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+    DUP2
+    // stack: len, b_loc, e_loc, m_loc, out, s1, s2, s3, s4, s5, expmod_contd, len, kexit_info
+
+    %jump(modexp_bignum)
+
+expmod_contd:
+    // stack: len, kexit_info
+
+    // Copy the result value from kernel general memory to the parent's return data.
+
+    DUP1
+    // stack: len, len, kexit_info
+    %mstore_parent_context_metadata(@CTX_METADATA_RETURNDATA_SIZE)
+    // stack: len, kexit_info
+    DUP1
+    // stack: len, len, kexit_info
+    %mul_const(3)
+    // stack: out=3*len, len, kexit_info
+    PUSH 0
+    // stack: i=0, cur_address=out, len, kexit_info
+expmod_store_loop:
```

Comment:
Should go through the loop in the inverse direction to store the result in big-endian form.

Reply:
Ah right, thanks. I changed it so it goes through in reverse order

Would it be a problem if len/16 > l_M? That should just mean storing extra zeroes in return data (since the actual result of the expmod will necessarily be less than m), going past returndatasize = l_M — I assume that's okay?

Reply:
No that's not an issue, as long as `returndatasize = l_M` we can write as many zeros as we want after that.